ddl/placement/bundle.go

```diff
@@ -320,36 +306,140 @@ func (b *Bundle) Tidy() error {
 		if err != nil {
 			return err
 		}
-		// Constraints.Add() will automatically avoid duplication
-		// if -engine=tiflash is added and there is only one constraint
-		// then it must be -engine=tiflash
-		// it is seen as an empty constraint, so merge it
-		if len(rule.Constraints) == 1 {
-			extraCnt[rule.Role] += rule.Count
+		rule.ID = strconv.Itoa(id)
+		tempRules = append(tempRules, rule)
+		id++
+	}
+
+	groups := make(map[string]*ConstraintsGroup)
+	finalRules := tempRules[:0]
+	for _, rule := range tempRules {
+		key := rule.Constraints.FingerPrint()
+		existing, ok := groups[key]
+		if !ok {
+			groups[key] = &ConstraintsGroup{rules: []*Rule{rule}}
 			continue
 		}
-		rule.ID = strconv.Itoa(i)
-		newRules = append(newRules, rule)
-	}
-	for role, cnt := range extraCnt {
-		// add -engine=tiflash, refer to tidb#22065.
-		newRules = append(newRules, &Rule{
-			ID:    string(role),
-			Role:  role,
-			Count: cnt,
-			Constraints: []Constraint{{
-				Op:     NotIn,
-				Key:    EngineLabelKey,
-				Values: []string{EngineLabelTiFlash},
-			}},
-			// the merged rule should have the same location labels with the original rules.
-			LocationLabels: locationLabels,
-		})
+		existing.rules = append(existing.rules, rule)
+	}
+	for _, group := range groups {
+		group.MergeRulesByRole()
+	}
+	if len(groups) == 1 {
+		for _, group := range groups {
+			group.MergeTransformableRoles()
+		}
+	} else {
+		if err := transformableLeaderConstraint(groups); err != nil {
+			return err
+		}
+	}
+	for _, group := range groups {
+		finalRules = append(finalRules, group.rules...)
+	}
+	// sort by id
+	sort.SliceStable(finalRules, func(i, j int) bool {
+		return finalRules[i].ID < finalRules[j].ID
+	})
+	b.Rules = finalRules
+	return nil
+}
+
+// ConstraintsGroup is a group of rules with the same constraints.
+type ConstraintsGroup struct {
+	rules           []*Rule
+	canBecameLeader bool
+	isLeaderGroup   bool
+}
+
+func transformableLeaderConstraint(groups map[string]*ConstraintsGroup) error {
+	var leaderGroup *ConstraintsGroup
+	canBecameLeaderNum := 0
+	for _, group := range groups {
+		if group.isLeaderGroup {
+			if !(leaderGroup == nil) {
+				return ErrInvalidPlacementOptions
+			}
+			leaderGroup = group
+		}
+		if group.canBecameLeader {
+			canBecameLeaderNum++
+		}
+	}
+	if leaderGroup != nil && canBecameLeaderNum == 1 {
+		leaderGroup.MergeTransformableRoles()
```

Comment:
Could you add some comments here or somewhere to explain why when two or more groups have a leader or voter role, we should skip `MergeTransformableRoles`?

Reply:
I agree. But if 1 leader and 2 followers in group1 and 2 voters in group2, can we merge group1 to voters?

Reply:
> I agree. But if 1 leader and 2 followers in group1 and 2 voters in group2, can we merge group1 to voters?

No. They don't share same constraints.

Reply:
> I agree. But if 1 leader and 2 followers in group1 and 2 voters in group2, can we merge group1 to voters?

that means the `group 1` has leader priority than group 2, so it's not transformable.